homeassistant/components/mqtt/mixins.py

```diff
@@ -1120,7 +1126,18 @@ def _setup_common_attributes_from_config(self, config: ConfigType) -> None:
             config.get(CONF_ENABLED_BY_DEFAULT)
         )
         self._attr_icon = config.get(CONF_ICON)
-        self._attr_name = config.get(CONF_NAME)
+        entity_name: str | None | UndefinedType = config.get(CONF_NAME, UNDEFINED)
+        # Only set _attr_name if it is needed
+        if entity_name is not UNDEFINED:
+            self._attr_name = entity_name
+        elif (
+            self._default_name is not UNDEFINED
+            and not self._default_to_device_class_name()
+        ):
+            self._attr_name = self._default_name
+        if CONF_DEVICE in config:
+            # Only set has_entity_name if a valid name is set
+            self._attr_has_entity_name = bool(config[CONF_DEVICE].get(CONF_NAME))
```

Comment:
Set `has_entity_name` to `False` if there is no valid device name.

Reply:
This will set it to True if a device name exists. We should only set it to False if it is not there.

Reply:
We should maybe also print a warning if we do so, to let user know what's going on ?